packages/dapi/lib/grpcServer/handlers/blockheaders-stream/getHistoricalBlockHeadersIteratorFactory.js

```diff
@@ -34,17 +36,45 @@ function getHistoricalBlockHeadersIteratorFactory(coreRpcApi) {
 
     for (let batchIndex = 0; batchIndex < numberOfBatches; batchIndex++) {
       const currentHeight = fromBlockHeight + batchIndex * MAX_HEADERS_PER_REQUEST;
-      const blocksToScan = getBlocksToScan(batchIndex, numberOfBatches, count);
 
-      const blockHash = await coreRpcApi.getBlockHash(currentHeight);
+      let blocksToScan = getBlocksToScan(batchIndex, numberOfBatches, count);
+      let blockHash = await coreRpcApi.getBlockHash(currentHeight);
+
+      const blockHeights = [...Array(blocksToScan).keys()]
+        .map((e, i) => currentHeight + i)
+
+      const cachedBlockHeaders = blockHeights.map((height) => cache.get(height))
+      const [firstCachedItem] = cachedBlockHeaders
+
+      let lastCachedIndex = -1
+
+      if (firstCachedItem) {
+        const firstMissingIndex = cachedBlockHeaders.indexOf(undefined)
+
+        // return cache if we do not miss anything
+        if (cachedBlockHeaders.filter(e => !!e).length === blocksToScan) {
+          return cachedBlockHeaders
+        }
+
+        if (firstMissingIndex !== -1) {
+          lastCachedIndex = firstMissingIndex - 1
+
+          const rawBlockHeader = cachedBlockHeaders[lastCachedIndex]
+          const blockHeader = BlockHeader.fromRawBlock(rawBlockHeader)
+
+          blockHash = blockHeader.hash.toString('hex')
```

Comment:
In this part, having we have some block headers cached, we take last cached block as a reference and count missing items for getBlockHeaders request. So, that means that getBlockHeaders will retrieve not just missing blocks, but also the first one, that we obviously already have in the DAPI.

The reason for that is to avoid additional call to the Dash Core node, because then only way to retrieve blockhash is to fire getBlockHash request which takes blockHeight as the argument.

What do you think guys?

Reply:
I think it's fine. There seems no other straightforward way to do that